### PR TITLE
Introducing custom blips: as easy as a custom showname!

### PR DESCRIPTION
**Pull request by Lernos**
Requires modifications from server owners for their servers to include 'custom_blips' in their "FL" packet and a new "validation monstrosity" to include a new string called blipname which must be added to the IC message args processed by the server. For backwards compatibility I suggest setting blipname as a server-side args variable to an empty string "" by default, since older clients will send nothing. In that case, the client will read the char.ini file to determine the correct blips/gender instead of looking it up in the new IC message packet.
P.S. Obviously themes need to be edited too, just add an ao2_custom_blips box somewhere near the ao2_ic_chat_name as needed.

### CODE REVIEW

**Comment by Lernos** [comment]
I know, clang-tidy, I don't love you either.

**Comment by in1tiate** [comment]
This does not need to be part of the IC validation hell string, and as a whole blips do not need to be networked. Voting for closure.

**Comment by Lernos** [comment]
Why do they not need to be networked though? We allow to play custom SFX on any message, but not blips? How about the fact that canonically, Godot can imitate voices of other characters, Ted Tonate can speak either as a "ddmale" or "tonate" typing on his keyboard, Robin changes from male to female voice after her transformation, if you want to have an invisible narrator ("Anon" type of character), you'd have to make one for **each** blipvoice? These things are all important to players I know. Changing a character's blips to any other blips on the fly seems like a very nice customization option.

**Comment by Salanto** [comment]
This is double edged.
It has terrible abuse potential, but at the same time is very flexible frame_sfx is already being used to earrape everyone around them and I doubt this would fare much better.

**Comment by in1tiate** [comment]
> Why do they not need to be networked though? We allow to play custom SFX on any message, but not blips? How about the fact that canonically, Godot can imitate voices of other characters, Ted Tonate can speak either as a "ddmale" or "tonate" typing on his keyboard, Robin changes from male to female voice after her transformation, if you want to have an invisible narrator ("Anon" type of character), you'd have to make one for **each** blipvoice? These things are all important to players I know. Changing a character's blips to any other blips on the fly seems like a very nice customization option.

I hadn't considered characters with multiple blip types, which is a feature the unofficial 2.5 client had and which we currently still lack. With that in mind I'm in favor of this change.

**Review by in1tiate — CHANGES_REQUESTED** [review]
See above; please avoid adding clutter to the main UI

**Comment by TrickyLeifa** [comment]
I'd recommend linking a server pull-request that introduces the network changes so that it's easier to test the changes; and vice versa.

File: src/courtroom.cpp [hunk 1]
```diff
@@ -2083,6 +2101,10 @@ void Courtroom::on_chat_return_pressed()
     }
   }
 
+  if (ao_app->custom_blips_enabled) {
+    if (ui_custom_blips->text().isEmpty()) { packet_contents.append(ui_custom_blips->placeholderText()); }
+    else { packet_contents.append(ui_custom_blips->text()); }
```
Comment (oldmud0) [inline src/courtroom.cpp:2106]: Please keep your if blocks unfolded

**Review by oldmud0 — COMMENTED** [review]
Looks fine but seems abusable - e.g. set blip name to `../music/zettaslow`.

Also looking for the serverside PR for this.

Also put that profile picture away.

File: src/courtroom.cpp [hunk 2]
```diff
@@ -177,6 +177,11 @@ Courtroom::Courtroom(AOApplication *p_ao_app) : QMainWindow()
   ui_ic_chat_name->setPlaceholderText(tr("Showname"));
   ui_ic_chat_name->setText(p_ao_app->get_default_showname());
   ui_ic_chat_name->setObjectName("ui_ic_chat_name");
+  
+  ui_custom_blips = new QLineEdit(this);
```
Comment (oldmud0) [inline src/courtroom.cpp:181]: Oh yeah also this definitely should be a dropdown. It's not ideal to just throw the player into the darkness and make them scroll through the files themselves.

**Comment by Salanto** [comment]
By the by, @Lernos. You could join the Discord to make communication a bit easier.

**Comment by Crystalwarrior** [comment]
> Looks fine but seems abusable - e.g. set blip name to `../music/zettaslow`.
> 
> Also looking for the serverside PR for this.
> 
> Also put that profile picture away.

Well servers should sanitize it in that case lmao

**Comment by Lernos** [comment]
> Also put that profile picture away.

Sorry xD

> By the by, @Lernos. You could join the Discord to make communication a bit easier.

Oh, I actually am. I was just dealing with stuff IRL. Not sure it would make communications that much easier, but whatever's more convenient.

> Also same, serverside PR would be nice

There you go. I don't work with akashi, but I'm pretty sure it can be figured out. There should be a similar mechanism there somewhere, right?
AttorneyOnline/tsuserver3#198

**Comment by Salanto** [comment]
I'll handle Akashi. You might want to resolve those merge conflicts tho.

**Comment by stonedDiscord** [comment]
> Oh yeah also this definitely should be a dropdown. It's not ideal to just throw the player into the darkness and make them scroll through the files themselves.

but which sounds should the dropdown contain?

1. create an` .ini` file that defines which one should be
2. every .opus file in `base\sounds\blips`
3. every file in `base\sounds\general` that starts with `sfx-blip`
4. every file from every subfolder in `base\sounds\blips`
5. all of the above

if we go with 5, what to do about the list getting too long?

**Comment by oldmud0** [comment]
I would go with 2 although it wouldn't be much additional effort to go with 4.

**Comment by in1tiate** [comment]
I plan on pulling the network changes from this pull request into the code I'm writing for #952 unless there are any objections. I will not be including the UI changes, as they are outside the scope of that issue.

**Comment by Salanto** [comment]
Stale PR + Implemented in a different PR

**Pull request closed without merging**